viskores/exec/SplineEvaluateRectilinearGrid.h

```diff
@@ -0,0 +1,259 @@
+//============================================================================
+//  The contents of this file are covered by the Viskores license. See
+//  LICENSE.txt for details.
+//
+//  By contributing to this file, all contributors agree to the Developer
+//  Certificate of Origin Version 1.1 (DCO 1.1) as stated in DCO.txt.
+//============================================================================
+
+//============================================================================
+//  Copyright (c) Kitware, Inc.
+//  All rights reserved.
+//  See LICENSE.txt for details.
+//
+//  This software is distributed WITHOUT ANY WARRANTY; without even
+//  the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
+//  PURPOSE.  See the above copyright notice for more information.
+//============================================================================
+#ifndef viskores_exec_SplineEvaluateRectilinearGrid_h
+#define viskores_exec_SplineEvaluateRectilinearGrid_h
+
+#include <viskores/Bounds.h>
+#include <viskores/ErrorCode.h>
+
+namespace viskores
+{
+namespace exec
+{
+
+class VISKORES_ALWAYS_EXPORT SplineEvaluateRectilinearGrid
+{
+private:
+  using FieldType = viskores::cont::ArrayHandle<viskores::FloatDefault>;
+  using AxisType = FieldType;
+
+  using FieldPortalType = typename FieldType::ReadPortalType;
+  using AxisPortalType = typename AxisType::ReadPortalType;
+
+public:
+  VISKORES_CONT SplineEvaluateRectilinearGrid() = default;
+
+  template <typename CoordsType, typename ArrayPortalType>
+  VISKORES_CONT SplineEvaluateRectilinearGrid(const CoordsType& coords,
+                                              const ArrayPortalType& field)
+    : Field(field)
+  {
+    this->AxisPortals[0] = coords.GetFirstPortal();
+    this->AxisPortals[1] = coords.GetSecondPortal();
+    this->AxisPortals[2] = coords.GetThirdPortal();
+
+    this->NumX = this->AxisPortals[0].GetNumberOfValues();
+    this->NumY = this->AxisPortals[1].GetNumberOfValues();
+    this->NumZ = this->AxisPortals[2].GetNumberOfValues();
+    this->Bounds = viskores::Bounds(this->AxisPortals[0].Get(0),
+                                    this->AxisPortals[0].Get(this->NumX - 1),
+                                    this->AxisPortals[1].Get(0),
+                                    this->AxisPortals[1].Get(this->NumY - 1),
+                                    this->AxisPortals[2].Get(0),
+                                    this->AxisPortals[2].Get(this->NumZ - 1));
```

Comment:
This code looks problematic to me. This constructor runs in the control environment, but `AxisPortals` points to data in the execution environment. I'm surprised you are not getting errors on Kokkos builds with separate host/device memory spaces.

I suggest changing this constructor to accept the control-side objects and call `PrepareForInput` on the arrays here instead of the caller.